internal/controller/clusterorder_controller.go

```diff
@@ -19,47 +19,261 @@ package controller
 
 import (
 	"context"
+	"fmt"
 
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	rand "k8s.io/apimachinery/pkg/util/rand"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
+	controllerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	cloudkitv1alpha1 "github.com/innabox/cloudkit-operator/api/v1alpha1"
 )
 
+// NewComponentFn is the type of a function that creates a required component
+type NewComponentFn func(context.Context, *cloudkitv1alpha1.ClusterOrder) (*appResource, error)
+
+type appResource struct {
+	object      client.Object
+	mutateFn    controllerutil.MutateFn
+	shouldExist bool
+}
+
+type component struct {
+	name string
```

Comment:
nit: `kind` instead of `name` would be more consistent terminology with the k8s API conventions.

Reply:
This really is meant to be a "name" and not a "kind"; it comes from [one of the examples you provided](https://github.com/saas-patterns/boutique-shop-operator/blob/6ad9b7112b15c14acdfc5dc0e8ac19aef066e7db/controllers/boutiqueshop_controller.go#L48). It's used as a human-readable descriptor in some log messages (and it makes the code a little easier to read to have the label there along with the function).

Reply:
Yeah, the guy who made that example probably should have called it `kind` too. ;)

Anyone looking at controller logs, or especially controller code, should know what Group, Version and Kind are. That's the official terminology for defining a k8s API type.

https://book.kubebuilder.io/cronjob-tutorial/gvks.html
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds

Reply:
That said, it's not critical. And potentially when we get to creating multiple resources of the same Kind, for example multiple ConfigMaps or Secrets would be common, then we'll perhaps put something in this field that's a combination of kind and some other designation to give context.